lib/modules/manager/sbt/extract.ts

```diff
@@ -357,9 +391,122 @@ export function extractPackageFile(content: string): PackageFile | null {
     isMultiDeps: false,
     scalaVersion: null,
     variables: {},
+    ...defaultAcc,
   };
 
   // TODO: needs major refactoring?
   const res = lines.reduce(parseSbtLine, acc);
   return res.deps.length ? res : null;
 }
+
+function prepareLoadPackageFiles(
+  _config: ExtractConfig,
+  packageFilesContent: MapFilenameContent
+): {
+  variables: ParseOptions['variables'];
+  registryUrls: string[];
+  scalaVersion: ParseOptions['scalaVersion'];
+} {
+  let variables: ParseOptions['variables'] = {};
+  const registryUrls: string[] = [MAVEN_REPO];
+  const acc: PackageFile & ParseOptions = {
+    registryUrls,
+    deps: [],
+    variables,
+  };
+  let scalaVersion: string | null = null;
+
+  for (const [packageFile, content] of Object.entries(packageFilesContent)) {
+    acc.packageFile = packageFile;
+    const res = extractFile(content, acc);
+    if (res) {
+      variables = { ...variables, ...res.variables };
+      if (res.registryUrls) {
+        registryUrls.push(...res.registryUrls);
+      }
+      if (res.scalaVersion) {
+        scalaVersion = res.scalaVersion;
+      }
+    }
+  }
+
+  return {
+    variables,
+    registryUrls,
+    scalaVersion,
+  };
+}
+
+export async function extractAllPackageFiles(
+  _config: ExtractConfig,
+  packageFiles: string[]
+): Promise<PackageFile[] | null> {
+  // Start parsing file in project/ folder first to get variable
+  packageFiles.sort((a, b) => (a.match('project/.*\\.scala$') ? -1 : 1));
```

Comment:
Thinking about multi-project repositories... 🤔

What if we do something like this:
1. Separate `project/*.scala` files from `build.sbt` ones
2. Iterate over files like `foo/build.sbt` and `bar/build.sbt`, first extracting dependencies, variables and registry URLs from `foo/project/*.scala` and `bar/project/*.scala` files
3. Extract `build.sbt` files using only variables and registry URLs extracted from appropriate `*.scala` files

Reply:
So you're suggesting ordering like this right

1. project/*.scala
2. others-submodule/project/*.scala
3. resolve all *.sbt file

Reply:
I think of resolving which `.scala` file belongs to which `.sbt` file first and make sure extracted variables aren't used anywhere else

Reply:
But there's a case where `submodule/build.sbt` is using variable from `project/Versions.scala` too.

1 `project/Versions.scala` file might belong to many submodule's `build.sbt`

Reply:
But i think i got your point now. 

`submodule/project/Versions.scala` can only be use within `submodule/build.sbt` right?
For `project/Versions.scala` can be use to all submodules and root `build.sbt`

Reply:
What I came up with is to have
1. `globalVariables` from root of repository and `project/`
2. local variable within the submodule

So it loops in `root` and `project/` first to get the global variable

then extract the package file by folder group

Reply:
@zharinov Please help review again

Reply:
use Minimatch instead of regex, as this should work on windows where we've backslash instead of forward slash as separator